gtk4/src/subclass/entry_buffer.rs

```diff
@@ -0,0 +1,242 @@
+    fn parent_deleted_text(&self, entry_buffer: &Self::Type, position: u32, n_chars: u32);
+    fn parent_get_length(&self, entry_buffer: &Self::Type) -> u32;
+    fn parent_get_text(&self, entry_buffer: &Self::Type) -> GString;
+    fn parent_insert_text(&self, entry_buffer: &Self::Type, position: u32, chars: &str) -> u32;
+    fn parent_inserted_text(&self, entry_buffer: &Self::Type, position: u32, chars: &str);
+}
+
+impl<T: EntryBufferImpl> EntryBufferImplExt for T {
+    fn parent_delete_text(&self, entry_buffer: &Self::Type, position: u32, n_chars: u32) -> u32 {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut ffi::GtkEntryBufferClass;
+            let f = (*parent_class)
+                .delete_text
+                .expect("No parent class impl for \"delete_text\"");
+            f(
+                entry_buffer
+                    .unsafe_cast_ref::<EntryBuffer>()
+                    .to_glib_none()
+                    .0,
+                position,
+                n_chars,
+            )
+        }
+    }
+
+    fn parent_deleted_text(&self, entry_buffer: &Self::Type, position: u32, n_chars: u32) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut ffi::GtkEntryBufferClass;
+            if let Some(f) = (*parent_class).deleted_text {
+                f(
+                    entry_buffer
+                        .unsafe_cast_ref::<EntryBuffer>()
+                        .to_glib_none()
+                        .0,
+                    position,
+                    n_chars,
+                )
+            }
+        }
+    }
+
+    fn parent_get_length(&self, entry_buffer: &Self::Type) -> u32 {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut ffi::GtkEntryBufferClass;
+            let f = (*parent_class)
+                .get_length
+                .expect("No parent class impl for \"get_length\"");
+            f(entry_buffer
+                .unsafe_cast_ref::<EntryBuffer>()
+                .to_glib_none()
+                .0)
+        }
+    }
+
+    fn parent_get_text(&self, entry_buffer: &Self::Type) -> GString {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut ffi::GtkEntryBufferClass;
+            let f = (*parent_class)
+                .get_text
+                .expect("No parent class impl for \"get_text\"");
+            let mut n_bytes = 0;
+            let res = f(
+                entry_buffer
+                    .unsafe_cast_ref::<EntryBuffer>()
+                    .to_glib_none()
+                    .0,
+                &mut n_bytes,
+            );
+            FromGlibContainer::from_glib_none_num(res, n_bytes as usize)
```

Comment:
This needs some more work AFAIU. What you get back from the C function is the number of *characters*, but for creating the sub-slice we need to number of bytes (you correctly name it `n_bytes` here).

You need to create a `&str` from the value returned by C (making use of the `NUL` terminator to never read too much) and then use the unicode APIs to know what byte position corresponds to the n-th character.

Reply:
Strangely, this method returns the number of bytes and not the number of characters like the `insert|delete{_ed}_text` ones

Reply:
Might be worth reporting this inconsistency to the GTK people :)

Reply:
Too late for GTK4 at this point

Reply:
I actually asked mclassen about this one before gtk4 release, it's intended for it to be n_bytes & not n_chars